src/crypto/tls/handshake_server_tls13.go

```diff
@@ -279,23 +280,38 @@ GroupSelection:
 		clientKeyShare = &hs.clientHello.keyShares[0]
 	}
 
-	if _, ok := curveForCurveID(selectedGroup); selectedGroup != X25519 && !ok {
+	if _, ok := curveForCurveID(selectedGroup); selectedGroup != X25519 && curveIdToCirclScheme(selectedGroup) == nil && !ok {
 		c.sendAlert(alertInternalError)
 		return errors.New("tls: CurvePreferences includes unsupported curve")
 	}
-	params, err := generateECDHEParameters(c.config.rand(), selectedGroup)
-	if err != nil {
-		c.sendAlert(alertInternalError)
-		return err
+	if kem := curveIdToCirclScheme(selectedGroup); kem != nil {
+		ct, ss, internalErr, err := encapsulateForKem(kem, c.config.rand(), clientKeyShare.data)
+		if err != nil {
+			if internalErr {
+				c.sendAlert(alertInternalError)
+			} else {
+				c.sendAlert(alertIllegalParameter)
+			}
+			return err
```

Comment:
Can we review all newly added error paths, and check whether the error messages (when hit) are sufficiently unique to recognize where something went wrong? And if not, maybe add more context?

For example, `encapsulateForKem` can return `internalErr == true` if reading from the randomness source fails, it could result in something like `EOF`. Presumably that is such a unique case that we can ignore.

However if the client provided some crap and `UnmarshalBinaryPublicKey` returns a failure, what kind of errors can we expect there? I think it would be something like `wrong size for public key` for `ErrPubKeySize`. If that is the case, can we use something like `fmt.Errorf("tls: kem encapsulate failed: %v", err)`

Reply:
Wrapped errors.